include/deal.II/matrix_free/mapping_info.h

```diff
@@ -446,7 +447,7 @@ namespace internal
       /**
        * The pointer to the underlying Mapping object.
        */
-      SmartPointer<const Mapping<dim>> mapping;
+      SmartPointer<const dealii::hp::MappingCollection<dim>> mapping;
```

Comment:
I really don't like this `SmartPointer` here. Would it be possible to eliminate it?

Reply:
This was introduced back in #8856 and for good reasons, as one should be able to query the information underlying the object. It technically goes through an internal class but we really must expose this information one way or the other. By the way, it is also this way in `FEValues`.

Thus, we should try a little harder than simply eliminate it (I see why you want to, though). My idea would be to utilize `Mapping::clone()` and store an `std::unique_ptr` with the mapping instead. While the type changes, all typical use cases would be equivalent as it is still a pointer type. Since some mappings have a relatively high memory consumption, we should look over the most important ones, say `MappingQGeneric`, and see if we can make some big members like the interpolation tables `std::shared_ptr` to share them via `MappingQGeneric::clone()`.

Reply:
@kronbichler I went with `std::shared_ptr<dealii::hp::MappingCollection<dim>>`.